exec/java-exec/src/main/java/org/apache/drill/exec/store/SchemaTreeProvider.java

```diff
@@ -119,6 +127,74 @@ public SchemaPlus createRootSchema(SchemaConfig schemaConfig) {
     }
   }
 
+
+  public SchemaPlus createPartialRootSchema(final String userName, final SchemaConfigInfoProvider provider,
+                                            final String storage) {
+    final String schemaUser = isImpersonationEnabled ? userName : ImpersonationUtil.getProcessUserName();
+    final SchemaConfig schemaConfig = SchemaConfig.newBuilder(schemaUser, provider).build();
+    final SchemaPlus rootSchema = SimpleCalciteSchema.createRootSchema(false);
+    Set<String> storageSet = Sets.newHashSet();
+    storageSet.add(storage);
+    addNewStoragesToRootSchema(schemaConfig, rootSchema, storageSet);
+    schemaTreesToClose.add(rootSchema);
+    return rootSchema;
+  }
+
+  public SchemaPlus addPartialRootSchema(final String userName, final SchemaConfigInfoProvider provider,
+                                            Set<String> storages, SchemaPlus rootSchema) {
+    final String schemaUser = isImpersonationEnabled ? userName : ImpersonationUtil.getProcessUserName();
+    final SchemaConfig schemaConfig = SchemaConfig.newBuilder(schemaUser, provider).build();
+    addNewStoragesToRootSchema(schemaConfig, rootSchema, storages);
+    schemaTreesToClose.add(rootSchema);
+    return rootSchema;
+  }
+
+  private void expandSecondLevelSchema(SchemaPlus parent) {
+    List<SchemaPlus> secondLevelSchemas = Lists.newArrayList();
+    for (String firstLevelSchemaName : parent.getSubSchemaNames()) {
+      SchemaPlus firstLevelSchema = parent.getSubSchema(firstLevelSchemaName);
+      for (String secondLevelSchemaName : firstLevelSchema.getSubSchemaNames()) {
+        secondLevelSchemas.add(firstLevelSchema.getSubSchema(secondLevelSchemaName));
+      }
+    }
+
+    for (SchemaPlus schema : secondLevelSchemas) {
+      AbstractSchema drillSchema;
+      try {
+        drillSchema = schema.unwrap(AbstractSchema.class);
+      } catch (ClassCastException e) {
+        throw new RuntimeException(String.format("Schema '%s' is not expected under root schema", schema.getName()));
+      }
+      SubSchemaWrapper wrapper = new SubSchemaWrapper(drillSchema);
+      parent.add(wrapper.getName(), wrapper);
+    }
+  }
+
+  public void addNewStoragesToRootSchema(SchemaConfig schemaConfig, SchemaPlus rootSchema, Set<String> storages) {
+    try {
+      for(String name: storages) {
+        try {
+          StoragePlugin plugin = dContext.getStorage().getPlugin(name);
+          if(plugin == null) {
+            logger.error("Got null for storage plugin of name: " + name);
+            continue;
+          }
+          plugin.registerSchemas(schemaConfig, rootSchema);
+          expandSecondLevelSchema(rootSchema);
+        }
+        catch (ExecutionSetupException ex) {
```

Comment:
Presumably this method adds storages based on need: we add a storage only if a query references it. If so, then we need to pass the exception along. If we don't, the user will simply be told that "foo.bar" does not exist when, in fact, perhaps it does exist but is configured wrong.

And, note that we do catch an IOException and translate it to a user exception below.

So, maybe

* Remove this try/catch block.
* Change the following catch from IOException to just Exception.

The result will be that we fail the query for all errors, not just IO errors. Again, this is the right choice if we are only loading the schemas needed by the query.